validator/keymanager/remote.go

```diff
@@ -0,0 +1,255 @@
+// Remote is a key manager that loads keys from a local Ethereum 2 wallet.
+type Remote struct {
+	paths    []string
+	conn     *grpc.ClientConn
+	accounts map[[48]byte]*accountInfo
+}
+
+type accountInfo struct {
+	Name   string `json:"name"`
+	PubKey []byte `json:"pubkey"`
+}
+
+type remoteOpts struct {
+	Location     string                 `json:"location"`
+	Accounts     []string               `json:"accounts"`
+	Certificates *remoteCertificateOpts `json:"certificates"`
+}
+
+type remoteCertificateOpts struct {
+	CACert     string `json:"ca_cert"`
+	ClientCert string `json:"client_cert"`
+	ClientKey  string `json:"client_key"`
+}
+
+var remoteOptsHelp = `The remote key manager connects to a walletd instance.  The options are:
+  - location This is the location to look for wallets.  If not supplied it will
+    use the standard (operating system-dependent) path.
+  - accounts This is a list of account specifiers.  An account specifier is of
+    the form <wallet name>/[account name],  where the account name can be a
+    regular expression.  If the account specifier is just <wallet name> all
+    accounts in that wallet will be used.  Multiple account specifiers can be
+    supplied if required.
+  - certificates This provides paths to certificates:
+    - ca_cert This is the path to the server's certificate authority certificate file
+    - client_cert This is the path to the client's certificate file
+    - client_key This is the path to the client's key file
+
+An sample keymanager options file (with annotations; these should be removed if
+using this as a template) is:
+
+  {
+	"location":    "host.example.com:12345", // Connect to walletd at host.example.com on port 12345
+    "accounts":    ["Validators/Account.*"]  // Use all accounts in the 'Validators' wallet starting with 'Account'
+	"certificates": {
+	  "ca_cert": "/home/eth2/certs/ca.crt"         // Certificate file for the CA that signed the server's certificate
+	  "client_cert": "/home/eth2/certs/client.crt" // Certificate file for this client
+	  "client_key": "/home/eth2/certs/client.key"  // Key file for this client
+	}
+  }`
+
+// NewRemote creates a key manager populated with the keys from walletd.
+func NewRemote(input string) (KeyManager, string, error) {
+	opts := &remoteOpts{}
+	err := json.Unmarshal([]byte(input), opts)
+	if err != nil {
+		return nil, remoteOptsHelp, err
+	}
+
+	if len(opts.Accounts) == 0 {
+		return nil, remoteOptsHelp, errors.New("at least one account specifier is required")
+	}
+
+	// Load the client certificates.
+	clientPair, err := tls.LoadX509KeyPair(opts.Certificates.ClientCert, opts.Certificates.ClientKey)
+	if err != nil {
+		return nil, remoteOptsHelp, errors.Wrap(err, "failed to obtain client's certificate and/or key")
+	}
+
+	// Load the CA for the server certificate.
+	serverCA, err := ioutil.ReadFile(opts.Certificates.CACert)
+	if err != nil {
+		return nil, remoteOptsHelp, errors.Wrap(err, "failed to obtain server's CA certificate")
```

Comment:
Absolutely.  Without mutual authentication the server misses a key piece of information to decide what actions the client is allowed to undertake.

Reply:
that seems to me the kind of logic that should be out of the scope of the validator.. that's in the realm of bearer keys, authentication and so on

Reply:
Existing Prysm RPC is already designed to work with certificate-based authentication with gRPC _e.g._ beacon node to validator communications.  This follows that pattern; if you want to change the general pattern it would be best to raise a separate issue to discuss the benefits of doing so.

Reply:
Just meant that it should be optional not mandatory. Validator checks if it was provided with a cert, if not then it creates an insecure connection.

The reason why I'm bringing this up is that a setup that I think people will go with is embedding a beacon node and a validator in something like k8 and then having a service which opens their rpc to the world. Forcing certs on the node/ validator level will make it harder. 
Flexibility is key when it comes to managing nodes/ validators in a more advanced setup.

Reply:
Given the sensitivity of signing operations, I think it makes sense to mandate client certificates.  We should not provide something that is insecure by default (for example, I've never seen anyone bother to put certificates between beacon node and validator even though it is possible to do so) when the majority of people running a remote signer will do so remotely (if local it would be far more efficient to run the local wallet keymanager).

Reply:
What if I embed several validators in a k8 setup and I want to manage authentication outside the validator? If we force the certs anyone with such a setup will need to change the validator code and rebuilt.
I suspect many will evolve to such a setup

Reply:
If you have a setup that complicated you should be able to generate a few client certificates.

The point here is to ensure that clients are as secure as possible by default.  Prysm uses gRPC, which uses client certificates for mutual authentication.  Creation of client certificates is a relatively painless piece of work, and we'll make sure that details as to how to do so are provided.  If a particular server implementation wants to ignore the information provided by the certificates it can, but that's not a reason to make client certificates optional as the vast majority of implementations will be able to use the client certificate as the authorisation mechanism.

Reply:
we can make it optional as it is for the validator itself.